File: JSLibrary/NeuralNetwork.js
class NeuralNetwork{

    constructor(input_nodes, hidden_nodes, output_nodes, learning_rate, startingbounds, activation_function) {
        
        this.learning_rate = learning_rate;
        
        //init weights and biases with random values
        this.weights_IH = new Matrix(hidden_nodes, input_nodes);
        this.weights_HO = new Matrix(output_nodes, hidden_nodes);
        this.weights_IH.randomize(startingbounds[0], startingbounds[1]);
        this.weights_HO.randomize(startingbounds[0], startingbounds[1]);

        this.bias_H = new Matrix(hidden_nodes, 1);
        this.bias_O = new Matrix(output_nodes, 1);
        this.bias_H.randomize(startingbounds[0], startingbounds[1]);
        this.bias_O.randomize(startingbounds[0], startingbounds[1]);

        this.activation_function = activation_function;    
    }

    //networks processes output for given input
    predict(input_array){

        //convert input array to vector
        let input = new Matrix(0, 0);
        input.fromArray(input_array);

        //generate hidden outputs
        let hidden_matrix = Matrix.multiply(this.weights_IH, input);
        hidden_matrix.add(this.bias_H);
        hidden_matrix.map(this.activation_function.func);

        //generate actual outputs
        let output_matrix = Matrix.multiply(this.weights_HO, hidden_matrix);
        output_matrix.add(this.bias_O);
        output_matrix.map(this.activation_function.func);

        let outputs = output_matrix.toArray();
        return outputs;
    }

    train(input_array, label_array){

        //get prediction for input from network
        let prediction_array = this.predict(input_array);
        let prediction = new Matrix(0, 0);
        prediction.fromArray(prediction_array);

        //convert label array into vector
        let label = new Matrix(0, 0);
        label.fromArray(label_array);

        //convert input array to vector
        let input = new Matrix(0, 0);
        input.fromArray(input_array);

        //get needed matrices: hidden matrix results (+ transposed), transposed weights, transposed input
        let hidden_matrix = Matrix.multiply(this.weights_IH, input);
        hidden_matrix.add(this.bias_H);
        hidden_matrix.map(this.activation_function.func);

        let input_t = Matrix.transpose(input);
        let hidden_matrix_t = Matrix.transpose(hidden_matrix);
        let weights_HO_t = Matrix.transpose(this.weights_HO);

        //get error vectors: errors proportioned by weights
        let output_error = Matrix.sub(label, prediction);
        let hidden_error = Matrix.multiply(weights_HO_t, output_error);

        //calculate gradients: 
        //learningrate *(scalar) errorvector *(hadamard) derivated activation function of output
        let output_gradients = Matrix.map(this.activation_function.dfunc, prediction);
        output_gradients.hadamard(output_error);
        output_gradients.scale(this.learning_rate);

        let hidden_gradients = Matrix.map(this.activation_function.dfunc, hidden_matrix);
        hidden_gradients.hadamard(hidden_error);
        hidden_gradients.scale(this.learning_rate);

        //calculate weight-deltas: gradients *(matrix multiplication) input ^ T
        let delta_weights_HO = Matrix.multiply(output_gradients, hidden_matrix_t);
        let delta_weights_IH = Matrix.multiply(hidden_gradients, input_t);

        //add gradients to biases and deltas to weights
        this.weights_HO.add(delta_weights_HO);
        this.bias_O.add(output_gradients);
        this.weights_IH.add(delta_weights_IH);
        this.bias_H.add(hidden_gradients);

        let errorrate = sqrt(output_error.absvalue())+sqrt(hidden_error.absvalue());
        return [prediction_array, errorrate];
    }   

}

File: JavaLibrary/NeuralNetwork.java
package JavaLibrary;

public class NeuralNetwork{

    private float learningRate;
    Matrix weigthsIH, weigthsHO, biasH, biasO;
    Function sig, derivSig;

    public NeuralNetwork(int inputNodes, int hiddenNodes, int outputNodes, float learningRate){
        this.learningRate = learningRate;

        //init weight- and bias-matrices with random values
        weigthsIH = new Matrix(hiddenNodes, inputNodes);
        weigthsIH.randomize(-1, 1);
        weigthsHO = new Matrix(outputNodes, hiddenNodes);
        weigthsHO.randomize(-1, 1);
        biasH = new Matrix(hiddenNodes, 1);
        biasH.randomize(-1, 1);
        biasO = new Matrix(outputNodes, 1);
        biasO.randomize(-1, 1);

        sig = new Sigmoid();
        derivSig = new DerivatedSigmoid();
    }

    //neural network process for given input, returns outputs (depended on number of outputs)
    public float[] feedforward(float[] inputs){

        //change input array into matrix
        Matrix inputVector = new Matrix(1, inputs.length);
        inputVector.toVectorColumn(inputs);

        //generate hidden outputs
        Matrix hiddenMatrix = Matrix.multiply(weigthsIH, inputVector);
        hiddenMatrix = Matrix.add(hiddenMatrix, biasH);
        hiddenMatrix = Matrix.map(hiddenMatrix, sig);

        //generate actual outputs
        Matrix outputMatrix = Matrix.multiply(weigthsHO, hiddenMatrix);
        outputMatrix = Matrix.add(outputMatrix, biasO);
        outputMatrix = Matrix.map(outputMatrix, sig);
       
        float[] outputs = outputMatrix.getColumn(0);
        return outputs;
    }

    public void train(int epoches, Datapoint[] trainingData){
        for(int i=0; i<epoches; i++){
            float errorsum = 0;
            for(Datapoint datapoint: trainingData){
            //for(int j=0; j<trainingData.length; j++){
                //trainPoint(trainingData[i].inputs, trainingData[i].labels); 
                errorsum += trainPoint(datapoint.inputs, datapoint.labels);
            }
            System.out.println(String.format("Epoche %d: %f errorrate", i, errorsum/trainingData.length));
        }
    }

    private float trainPoint(float[] inputs, float[] labels){
        //test current solution provided by network, convert into matrix
        float[] guess = feedforward(inputs);
        Matrix guessVector = new Matrix(1, guess.length);
        guessVector.toVectorColumn(guess);

        //convert labels into matrix
        Matrix labelVector = new Matrix(1, labels.length);
        labelVector.toVectorColumn(labels);

        //calculate output errors based on networks solution und actual solution
        Matrix outputErrorVector = Matrix.subtract(labelVector, guessVector);

        //calculate hidden errors based on the output errors (errors proportioned by the respective weights)
        //formula: weights^T * errorvector
        Matrix transposedWeights = Matrix.transpose(weigthsHO);
        Matrix hiddenErrorVector = Matrix.multiply(transposedWeights, outputErrorVector);

        //get the transposed inputvector and the transposed hiddenmatrix output
        Matrix inputVector = new Matrix(1, inputs.length);
        inputVector.toVectorColumn(inputs);
        Matrix transposedInputVector = Matrix.transpose(inputVector);
        Matrix hiddenMatrix = Matrix.multiply(weigthsIH, inputVector);
        hiddenMatrix = Matrix.add(hiddenMatrix, biasH);
        hiddenMatrix = Matrix.map(hiddenMatrix, sig);
        Matrix transposedHiddenMatrix = Matrix.transpose(hiddenMatrix);

        //calculate differences for the weights to be applied to correct the weights
        //formula : learningrate *(scalar) errorvector *(hadamard) derivated activation function of output *(matrix multiplication) inputvector^T
        //add the deltas to the weights
        //add the deltas without the transposed inputvectors to the biases to adjust those
        Matrix gradients = Matrix.map(guessVector, derivSig); 
        gradients = Matrix.hadamard_product(gradients, outputErrorVector);
        gradients = Matrix.scale(gradients, learningRate);
        Matrix delta_weightsHO = Matrix.multiply(gradients, transposedHiddenMatrix);
        biasO = Matrix.add(biasO, gradients);
        weigthsHO = Matrix.add(weigthsHO, delta_weightsHO);

        Matrix hiddenGradients = Matrix.map(hiddenMatrix, derivSig); 
        hiddenGradients = Matrix.hadamard_product(hiddenGradients, hiddenErrorVector);
        hiddenGradients = Matrix.scale(hiddenGradients, learningRate);
        Matrix delta_weightsIH = Matrix.multiply(hiddenGradients, transposedInputVector);
        biasH = Matrix.add(biasH, hiddenGradients);
        weigthsIH = Matrix.add(weigthsIH, delta_weightsIH);

        //return error rate probably
        return hiddenErrorVector.sumAbs()/hiddenErrorVector.m + outputErrorVector.sumAbs()/outputErrorVector.m;
    }

}